mammooc-course-evaluation-form.html

```diff
@@ -106,132 +114,150 @@
+            MammoocLocalizeBehavior,
+            Polymer.IronJsonpLibraryBehavior
+        ], Polymer.Element) {
+
+            static get is() {
+                return 'mammooc-course-evaluation-form';
+            }
+
+            static get properties() {
+                return {
+                    /**
+                     *    The provider of the course, e.g. "openHPI"
+                     */
+                    provider: {
+                        type: String
+                    },
+
+                    /**
+                     *   The ID of the course in the provider's system
+                     */
+                    courseId: {
+                        type: String
+                    },
+
+                    /**
+                     *   A boolean stating whether the user is logged in on mammooc
+                     */
+                    loggedIn: {
+                        type: Boolean,
+                        value: false,
+                        readOnly: true
+                    },
+
+                    _tries: {
+                        type: Number,
+                        value: 0
+                    }
+                }
+            }
 
+            static get observers() {
+                return [
+                ]
+            }
+
+            ready() {
+                this.addEventListener('api-load', this._apiLoaded.bind(this));
+                this.addEventListener('post-sent', this._postSent.bind(this));
+                // Every time the window gets focus, send a new request in case of changes on mammooc
+                // _tries is needed because the jsonp library won't resend if the url is the same
+                let e = this;
+                window.onfocus = function() {
+                    e.$.evaluationRequest.libraryUrl = e._generateUrl(e.provider, e.courseId) + '&try=' + e._tries;
+                    e._tries += 1;
+                };
+                super.ready()
+            }
 
-    Polymer({
-        is: 'mammooc-course-evaluation-form',
-        behaviors: [
-            MammoocLocalizeBehavior
-            ],
-        ready: function() {
-          // Every time the window gets focus, send a new request in case of changes on mammooc
-          // _tries is needed because the jsonp library won't resend if the url is the same
-            var e = this;
-            window.onfocus = function() {
-                e.$.evaluationRequest.libraryUrl = e._generateUrl(e.provider, e.courseId) + '&try=' + e._tries;
-                e._tries += 1;
-            };
-        },
-
-        _postSent: function(event) {
-            document.getElementById('evaluationsXHR').generateRequest();
-            this.$.successToast.open();
-        },
-
-        _submitForm: function() {
+            _postSent(event) {
+                this.parentNode.parentNode.parentNode.querySelector('#evaluationsXHR').generateRequest();
```

Comment:
Hmm, this looks like it is very closely coupled to the current HTML structure. When somebody introduces a new HTML element in the tree or changes the structure slightly, this will just break...

Reply:
You're absolutely right and I do not prefer the current solution. Unfortunately, I was not able yet to find another. Here's why - maybe you have an idea?:

`this`:
`<mammooc-course-evaluation-form>`

`this.parentNode`:
`<div class="card-actions">`

`this.parentNode.parentNode`:
`<paper-card elevation="1" aria-label="">`

`this.parentNode.parentNode.parentNode`:
`#shadow-root` (of the `mammooc-rating-widget`)

`this.parentNode.parentNode.parentNode.querySelector('#evaluationsXHR')`:
`<iron-ajax id="evaluationsXHR" auto="" hidden="">`

An alternative would be to use 
`document.querySelector('mammooc-rating-widget').shadowRoot.querySelector('#evaluationsXHR')` which leads to problems if multiple `mammooc-rating-widget`s are defined on a single page.

Unfortunately, I was not able to find a way using an "upwards" querySelector. `closest()`, for example, could be used to get the `paper-card`, but it doesn't the `mammooc-rating-widget` nor the corresponding shadow root.

Reply:
@MrSerth Can't you give the `<mammooc-rating-widget>` element a method that returns the `evaluationsXHR` request - or better, encapsulates sending the request? Then, you only need to (dynamically) look upwards in the DOM for the next `mammooc-rating-widget` element?